src/examples/mnist/MNIST_SP.cpp

```diff
@@ -107,110 +88,97 @@ vector<UInt*> read_mnist_images(string path) {
+        file.read( (char*) data_raw.data(), img_size);
         // Copy the data into an array of UInt's
-        auto data = new UInt[2 * img_size];
+        image_t data(img_size);
         // Apply a threshold to the image, yielding a B & W image.
-        for(UInt pixel = 0; pixel < img_size; pixel++) {
-            data[2 * pixel] = data_raw[pixel] >= 128 ? 1 : 0;
-            data[2 * pixel + 1] = 1 - data[2 * pixel];
+        for(const auto pixel: data_raw) {
+            data.push_back(pixel >= 128 ? 1u : 0u); 
         }
         retval.push_back(data);
-        delete[] data_raw;
     }
     return retval;
 }
+*/
+class MNIST {
 
+  private:
+    SpatialPooler sp;
+    SDR input;
+    SDR columns;
+    SDRClassifier clsr;
 
-int main(int argc, char **argv) {
-  UInt verbosity = 1;
-  int train_dataset_iterations = 1;
-  int opt;
-  while ( (opt = getopt(argc, argv, "tv")) != -1 ) {  // for each option...
-    switch ( opt ) {
-      case 't':
-          train_dataset_iterations += 1;
-        break;
-      case 'v':
-          verbosity = 1;
-        break;
-      case '?':
-          cerr << "Unknown option: '" << char(optopt) << "'!" << endl;
-        break;
-    }
-  }
-  UInt train_time = train_dataset_iterations * 60000;
+    UInt verbosity = 1;
+    const UInt train_dataset_iterations = 1u;
 
-  signal(SIGINT, keyboard_interrupt_handler);   // install stack trace printout on error
-  signal(SIGSEGV, keyboard_interrupt_handler);   // install stack trace printout on error
+  public:
 
-  SDR input({28, 28, 2});
-  SpatialPooler sp(
+void setup() {
+
+  input.initialize({28, 28, 1});
+  sp.initialize(
     /* numInputs */                    input.dimensions,
-    /* numColumns */                   {10, 10, 120},
-    /* potentialRadius */              0,  // hardcoded elsewhere
-    /* potentialPct */                 .0000001, // hardcoded elsewhere
-    /* globalInhibition */             true,
-    /* localAreaDensity */             .015,
-    /* numActiveColumnsPerInhArea */   -1,
-    /* stimulusThreshold */             28,
-    /* synPermInactiveDec */           .00928,
-    /* synPermActiveInc */             .032,
-    /* synPermConnected */             .422,
-    /* minPctOverlapDutyCycles */      0.,
-    /* dutyCyclePeriod */              1402,
-    /* boostStrength */                0,
-    /* CPP SP seed */                  0,
-    /* spVerbosity */                  verbosity,
-    /* wrapAround */                   0 // discarded
+    /* numColumns */                   {7, 7, 8}
     );
+  sp.setGlobalInhibition(false);
+  // mapping to input
+  sp.setPotentialRadius(5); //receptive field (hyper-cube) how each col maps to input
+  sp.setPotentialPct(0.5); //of the recept. field above, how much % of inputs individual col connects to?
+  sp.setStimulusThreshold(0); //FIXME no learning if this > 0
+  sp.setSynPermConnected(0.4);
+  sp.setSynPermActiveInc(0.2);
+  sp.setSynPermInactiveDec(0.02);
```

Comment:
@ctrl-z-9000-times how did you optimize the parameters (by hand?) I get awful results with default SP (11%, slightly above random)

Reply:
I wrote a program to optimize the parameters.  I think i used swarming?

Reply:
Thanks! I'll have to figure that too. 
Using yours original params, or some "hand-tuned" I scored pretty bad results. Though the "topological pooler" should have been faster, but performance of local inhibition should not be any worse.

Reply:
Some major differences in parameters from mine to yours:
* This has too few mini-columns 7x7x8=392, I think you will need at least 10x as many.
* The  sparsity should probably be closer to 1% than 5%.
* The permanence increment looks too large.